File: WeLegends/app/src/main/java/es/coru/andiag/welegends/common/mvp/base/BasePresenter.java
package es.coru.andiag.welegends.common.mvp.base;

import android.content.Context;

/**
 * Created by iagoc on 10/12/2016.
 */

public abstract class BasePresenter<V extends Context & BaseView> {
    private final static String TAG = BasePresenter.class.getSimpleName();

    private V mView;

    public V getView() {
        return mView;
    }

    public final void attach(V view) {
        this.mView = view;
        onViewAttached();
    }

    public final void detach() {
        this.mView = null;
        onViewDetached();
    }

    public abstract void onViewAttached();

    public abstract void onViewDetached();

}

File: WeLegends/app/src/main/java/es/coru/andiag/welegends/common/mvp/base/BasePresenter.kt
package es.coru.andiag.welegends.common.mvp.base

import android.content.Context

/**
 * Created by iagocanalejas on 10/12/2016.
 */
abstract class BasePresenter<V> where V : BaseView, V : Context {

    private val TAG = BasePresenter::class.java.simpleName

    var mView: V? = null
        get private set

    open fun attach(view: V) {
        this.mView = view
        onViewAttached()
    }

    open fun detach() {
        this.mView = null
        onViewDetached()
    }

    abstract fun onViewAttached()
    abstract fun onViewDetached()

}